src/inmanta/deploy/state.py

```diff
@@ -125,20 +184,89 @@ def reset(self) -> None:
         self.resource_state.clear()
         self.types_per_agent.clear()
 
+    def block_resource(self, resource: ResourceIdStr, details: ResourceDetails, transient: bool) -> None:
+        """
+        Mark the given resource as blocked, i.e. it's not deployable. This method updates the resource details
+        and the resource_status.
+
+        :param resource: The resource that should be blocked.
+        :param details: The details of the resource that should be blocked.
+        :param transient: True iff the given resource is blocked transitively. It's blocked because one of its dependencies
+                          is blocked.
+        """
+        self.resources[resource] = details
+        if resource in self.resource_state:
+            if not transient:
+                self.resource_state[resource].status = ResourceStatus.UNDEFINED
+            self.resource_state[resource].blocked = BlockedStatus.YES
+        else:
+            resource_status = ResourceStatus.UNDEFINED if not transient else ResourceStatus.HAS_UPDATE
+            self.resource_state[resource] = ResourceState(
+                status=resource_status, deployment_result=DeploymentResult.NEW, blocked=BlockedStatus.YES
+            )
+            self.types_per_agent[details.id.agent_name][details.id.entity_type] += 1
+        self.dirty.discard(resource)
+
+    def block_provides(self, resources: Set[ResourceIdStr]) -> set[ResourceIdStr]:
+        """
+        Marks the provides of the given resources as blocked.
+
+        Must be called under the scheduler lock. This method assumes that all the resources of the model version
+        are populated in the resources dictionary.
+
+        :param resources: The set of resources for which the provides have to be blocked.
+        :return: The set of dependent resources that were marked as blocked (transitively).
+        """
+        result = set()
+        for dependent_resource in self.requires.get_all_provides_transitively(resources):
```

Comment:
if may become relevant in the future to only traverse up the provides relation until we find one that is already blocked, but this will do for now I think.